File: cpp/LongestSubstringWithoutRepeatingCharacters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int start = 0;
        unordered_map<char, int> map;
        int result = 0;
        for (int i = 0; i < (int)s.length(); i++) {
            if (map.find(s.at(i)) != map.end()) {
                int temp = map[s.at(i)];
                while (start <= temp) {
                    map.erase(s.at(start));
                    start ++;
                }
            }
            map[s.at(i)] = i;
            result = max(result, i - start + 1);
        }
        return result;
    }
};

File: java/LongestSubstringWithoutRepeatingCharacters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        int slow = 0;
        int fast = 0;
        int result = 0;
        while (fast < s.length()) {
            if (map.containsKey(s.charAt(fast)) == true) {
                result = Math.max(fast-slow, result);
                int temp = map.get(s.charAt(fast));
                while (slow <= temp) {
                    map.remove(s.charAt(slow));
                    slow += 1;
                }
            }
            map.put(s.charAt(fast), fast);
            fast += 1;
        }
        result = Math.max(fast-slow, result);
        return result;
    }
}
